pwndbg/commands/heap.py

```diff
@@ -1,46 +1,127 @@
-    gdb.execute('call brk(%i)' % n)
+from pwndbg.color import bold, yellow, red, underline
+
+PREV_INUSE = 1
+IS_MMAPED = 2
+NON_MAIN_ARENA = 4
+
+def get_main_arena(addr=None):
+    if addr == None:
+        main_arena = gdb.lookup_symbol('main_arena')[0].value()
+    else:
+        if isinstance(addr, (long, int)):
+            addr = hex(addr)
+        main_arena = gdb.parse_and_eval('(struct malloc_state)*' + addr)
 
+    if main_arena == None:
+        print(red('Symbol \'main_arena\' not found. Try installing libc ' \
+                  'debugging symbols or specifying the main arena address ' \
+                  'and try again'))
+
+    return main_arena
+
+@pwndbg.commands.Command
 @pwndbg.commands.OnlyWhenRunning
-@pwndbg.commands.ParsedCommand
-def sbrk(n=0):
-    '''Get the address of sbrk(n=0)'''
-    gdb.execute('call sbrk(%i)' % n)
+def heap(addr=None):
+    main_arena = get_main_arena(addr)
+    if main_arena == None:
+        return
 
+    heap_base = None
+    for m in pwndbg.vmmap.get():
+        if m.objfile == '[heap]':
+            heap_base = m.vaddr
 
+    if heap_base == None:
+        print(red('Could not find the heap'))
+        return
+
+    top = main_arena['top']
+    last_remainder = main_arena['last_remainder']
 
+    print(bold('Top Chunk: ') + str(top))
+    print(bold('Last Remainder: ') + str(last_remainder))
+    print()
 
-p = argparse.ArgumentParser(prog='hheap')
+    # Print out all chunks on the heap
+    # TODO: Add an option to only print out free/allocated chunks
+    addr = heap_base
+    while addr <= top:
+        chunk = malloc_chunk(addr)
+        size = int(chunk['size'])
 
-p.add_argument('--size',
-               help='Heap size.  May be expressed as an integer or range (e.g. 32-64).')
-p.add_argument('--verbose', action='store_true',
-               help='Print more information')
-p.add_argument('--free', action='store_true',
-               help='Only show free slots')
-p.add_argument('address', type=int, default=0,
-               help='Heap allocation to display')
+        # Clear the bottom 3 bits
+        size &= ~7
+        addr += size
 
+@pwndbg.commands.Command
 @pwndbg.commands.OnlyWhenRunning
+def arena(addr=None):
+    main_arena = get_main_arena(addr)
+    if main_arena == None:
+        return
+
+    print(main_arena)
+
 @pwndbg.commands.Command
-def hheap(*a):
-    """Prints out heap information.
-    """ + p.format_help()
-    try:
-        args = p.parse_args(a)
-    except SystemExit:
+@pwndbg.commands.OnlyWhenRunning
+def bins(addr=None):
+    main_arena = get_main_arena(addr)
+    if main_arena == None:
         return
 
+    fastbins = main_arena['fastbinsY']
+    bins = main_arena['bins']
+
+    num_fastbins = int(fastbins.type.sizeof / fastbins.type.target().sizeof)
+    num_bins = int(bins.type.sizeof / bins.type.target().sizeof)
+    fd_field_offset = 16
+
+    print(underline(yellow('Fastbins')))
+    for i in range(num_fastbins):
```

Comment:
IIRC each bin corresponds to a particular allocation size (8/16/24/36 etc) in bytes